pkgs/servers/http/nginx/nix-etag-1.16.0.patch

```diff
@@ -0,0 +1,163 @@
+     if (!clcf->etag) {
+         return NGX_OK;
+     }
++    if (r->headers_out.last_modified_time == 1) {
++        real = ngx_realpath(path->data, NULL);
++
++        if (real == NULL) {
++            return NGX_OK;
++        }
++    }
+ 
+     etag = ngx_list_push(&r->headers_out.headers);
+     if (etag == NULL) {
+@@ -1598,16 +1606,52 @@ ngx_http_set_etag(ngx_http_request_t *r)
+     etag->hash = 1;
+     ngx_str_set(&etag->key, "ETag");
+ 
+-    etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
+-    if (etag->value.data == NULL) {
+-        etag->hash = 0;
+-        return NGX_ERROR;
+-    }
++    #define NIX_STORE_DIR "@nixStoreDir@"
++    #define NIX_STORE_LEN @nixStoreDirLen@
++
++    if (r->headers_out.last_modified_time == 1 && real != NULL
++        && !ngx_strncmp(real, NIX_STORE_DIR, NIX_STORE_LEN)
++        && real[NIX_STORE_LEN] == '/'
++        && real[NIX_STORE_LEN + 1] != '\0')
++    {
++        ptr1 = real + NIX_STORE_LEN;
++        *ptr1 = '"';
++
++        ptr2 = (u_char *) ngx_strchr(ptr1, '-');
++
++        if (ptr2 == NULL) {
++            ngx_free(real);
++            etag->hash = 0;
++            return NGX_ERROR;
++        }
+ 
+-    etag->value.len = ngx_sprintf(etag->value.data, "\"%xT-%xO\"",
+-                                  r->headers_out.last_modified_time,
+-                                  r->headers_out.content_length_n)
+-                      - etag->value.data;
++        *ptr2++ = '"';
++        *ptr2 = '\0';
++
++        etag->value.len = ngx_strlen(ptr1);
++        etag->value.data = ngx_pnalloc(r->pool, etag->value.len);
++
++        if (etag->value.data == NULL) {
++            ngx_free(real);
++            etag->hash = 0;
++            return NGX_ERROR;
++        }
++
++        ngx_memcpy(etag->value.data, ptr1, etag->value.len);
++        ngx_free(real);
++    } else {
++        etag->value.data = ngx_pnalloc(r->pool, NGX_OFF_T_LEN + NGX_TIME_T_LEN + 3);
++
++        if (etag->value.data == NULL) {
++            etag->hash = 0;
++            return NGX_ERROR;
++        }
++
++        etag->value.len = ngx_sprintf(etag->value.data, "\"%xT-%xO\"",
++                                      r->headers_out.last_modified_time,
++                                      r->headers_out.content_length_n)
++                          - etag->value.data;
++    }
+ 
+     r->headers_out.etag = etag;
+ 
+diff --git a/src/http/ngx_http_core_module.h b/src/http/ngx_http_core_module.h
+index 85f6d66d..a1f9fc10 100644
+--- a/src/http/ngx_http_core_module.h
++++ b/src/http/ngx_http_core_module.h
+@@ -485,7 +485,7 @@ ngx_int_t ngx_http_core_content_phase(ngx_http_request_t *r,
+ void *ngx_http_test_content_type(ngx_http_request_t *r, ngx_hash_t *types_hash);
+ ngx_int_t ngx_http_set_content_type(ngx_http_request_t *r);
+ void ngx_http_set_exten(ngx_http_request_t *r);
+-ngx_int_t ngx_http_set_etag(ngx_http_request_t *r);
++ngx_int_t ngx_http_set_etag(ngx_http_request_t *r, ngx_str_t *path);
```

Comment:
This is a bit unfortunate here, because this changes the API for third-party extension modules.

Reply:
Same concern, I don't think we should ever change API of any programs that we patch (i.e. no headers changes).